sw/device/tests/kmac_kmac_key_sideload_test.c

```diff
@@ -0,0 +1,217 @@
+  uint32_t max_key_version;
+  CHECK_STATUS_OK(
+      keymgr_testutils_max_key_version_get(keymgr, &max_key_version));
+
+  if (sideload_params.version > max_key_version) {
+    LOG_INFO("Key version %d is greater than the maximum key version %d",
+             sideload_params.version, max_key_version);
+    LOG_INFO("Setting key version to the maximum key version %d",
+             max_key_version);
+    sideload_params.version = max_key_version;
+  }
+
+  CHECK_STATUS_OK(
+      keymgr_testutils_generate_versioned_key(keymgr, sideload_params));
+  LOG_INFO("Keymgr generated HW output for Kmac");
+  uint32_t output_sideload[ARRAYSIZE(kKmacModes)][kKmacOutputLen];
+
+  for (size_t it = 0; it < ARRAYSIZE(kKmacModes); it++) {
+    CHECK_STATUS_OK(kmac_testutils_kmac(
+        kmac, kKmacModes[it], &kSoftwareKey, kCustomString, kCustomStringLen,
+        kKmacMessage, kKmacMessageLen, kKmacOutputLen, output_sideload[it],
+        /*capacity=*/NULL));
+  }
+  LOG_INFO("Computed KMAC output for sideloaded key.");
+
+  if (!test_phase) {
+    // In test phase 0, store the KMAC digest into the retention SRAM.
+    for (size_t it = 0; it < ARRAYSIZE(kKmacModes); it++) {
+      // Use kRetSramOwnerAddr + 4 to avoid overwriting test_phase_cnt.
+      sram_ctrl_testutils_write(
+          kRetSramOwnerAddr + 4 + 4 * it * kKmacOutputLen,
+          (sram_ctrl_testutils_data_t){.words = output_sideload[it],
+                                       .len = kKmacOutputLen});
+    }
+
+    LOG_INFO("Clearing the sideloaded key.");
+
+    // Enable "clear the key" toggle, so that previous sideload key port is
+    // cleared.
+    CHECK_DIF_OK(
+        dif_keymgr_sideload_clear_set_enabled(keymgr, kDifToggleEnabled));
+
+    // Disable "clear the key" toggle, so that the sideload key port is stable.
+    // Otherwise, the sideload port is continuously overwritten by fresh
+    // randomness every clock cycle.
+    CHECK_DIF_OK(
+        dif_keymgr_sideload_clear_set_enabled(keymgr, kDifToggleDisabled));
+
+    uint32_t output_sideload_bad[kKmacOutputLen];
+    // Initialize this output array because the following `kmac_testutils_kmac`
+    // function call fails early and is expected to *not* overwrite this array.
+    for (size_t i = 0; i < kKmacOutputLen; i++) {
+      output_sideload_bad[i] = i;
+    }
+
+    CHECK(kFailedPrecondition ==
+          status_err(kmac_testutils_kmac(kmac, kKmacModes[0], &kSoftwareKey,
+                                         kCustomString, kCustomStringLen,
+                                         kKmacMessage, kKmacMessageLen,
+                                         kKmacOutputLen, output_sideload_bad,
+                                         /*capacity=*/NULL)));
+    LOG_INFO(
+        "Ran KMAC with an invalid sideload key and checked that it fails.");
+
+    // Verify that the output array did not get overwritten.
+    for (size_t i = 0; i < kKmacOutputLen; i++) {
+      CHECK(output_sideload_bad[i] == i);
```

Comment:
I think the test implementation currently doesn't fully match the testplan and the description above.
- IIUC, we check here that if the sideload key is invalid, KMAC doesn't produce an output and instead reports an error. 
- The testplan specifies we should check that the KMAC cannot produce the same output as before. The description above states the same (use a software key and ensure the output doesn't match the sideload key).

These are two different things. Maybe we should check both. But what is important is that test and testplan are aligned.

Reply:
Yes, you  are right. Currently, I am only testing whether after clearing the sideload key, KMAC does not produce and output.

As this is not specified in the test description:
```
Clear sideload key from keymgr.
Verify that KMAC can no longer reproduce the same KMAC output.
```
What should I do between these two steps? Create a new sideload key with a different salt and check whether the output does not match? Or should I switch from a sideload key to a software key and do the comparison?

Reply:
I thought of using a software key but using the sideload key with different salt sounds like a better idea to me! Do you think that can be done easily or would it be a lot of work?

Reply:
Done, I've also adapted the test plan description accordingly.

Testing whether after clearing the sideload key KMAC is no longer able to produce a valid output is actually already tested in `keymgr_sideload_kmac_test`.

Reply:
Thanks @nasahlpa , this looks good now!